R/AlgorithmHandler.r

```diff
@@ -0,0 +1,143 @@
+            "pipe.r:getResponseAsJsonString",
+          error_type = "AlgorithmError"
+        )
+      ))
+    })
+  }
+runLoad_ <- function(onLoadMethod) {
+  state <- onLoadMethod()
+  print("PIPE_INIT_COMPLETE")
+  flush.console()
+  state
+}
+
+
+AlgorithmHandler <- methods::setRefClass(
+  "AlgorithmHandler",
+  fields = list(
+    applyMethod = "function",
+    onLoadMethod = "function",
+    pipeName = "character"
+  ),
+  methods = list(
+    serve=function() {
+      getInputData_ <- function(input) {
+        if (input$content_type == "binary") {
+          base64enc::base64decode(input$data)
+        } else {
+          input$data
+        }
+      }
+      # Begin startup
+      outputFile <- fifo("/tmp/algoout", open="w", blocking=TRUE)
+      inputFile <- file(pipeName)
+      open(inputFile)
+      loadResult <- tryCatch({
+        stage <- "loading"
+        state <- runLoad_(onLoadMethod)
+        list(state = state)
+      },
+      error = function(e) {
+        message <- toString(e)
+        formatted <-
+          gsub("\n", "", message)
+        list(error = list(
+          message = formatted,
+          stacktrace = stage,
+          error_type = "AlgorithmError"
+        ))
+      })
+      
+      #Finished loading, check if we failed and if not - start main algorithm loop
+      if (is.null(loadResult$error)) {
+        state <- loadResult$state
+        while (length(line <-
+                      readLines(inputFile, n = 1)) > 0) {
+          stage <- "parsing"
+          output <- tryCatch({
+            input <- rjson::fromJSON(line)
+            inputData <-
+              getInputData_(input)
+            stage <- "algorithm"
+            if (is.null(state)) {
+              output <- applyMethod(inputData)
+            } else {
+              output <- applyMethod(inputData, state)
+            }
+            getResponseObject_(output)
+          },
+          error = function(e) {
+            message <- toString(e)
+            formatted <-
+              gsub("\n", "", message)
+            list(error = list(
+              message = formatted,
+              stacktrace = stage,
+              error_type = "AlgorithmError"
+            ))
+          })
+          
+          # Flush stdout before writing back response
+          flush.console()
+          
+          response = getResponseAsJsonString_(output)
+          writeLines(response, con = outputFile)
+          # Finished writing response to algoout, checking stdin for more input.
+        }
+      } else{
+        # Flush stdout before writing back response
+        flush.console()
+        response = getResponseAsJsonString_(result)
+        writeLines(response, con = outputFile)
+      }
+      # Either the stdin pipe was closed, or we recieved in error. In any case we should shut down now
+      # and close the pipes.
+      close(inputFile)
+      close(outputFile)
+    }
+  )
+)
+
+
+getAlgorithmHandler <-
+  function(applyfunc,
```

Comment:
Can we put the function signature all on 1 line? The default onLoadMethod can be a floating function in this file to make that a bit easier to read